contracts/0.4.24/Lido.sol

```diff
@@ -651,6 +673,156 @@ contract Lido is ILido, StETH, AragonApp {
+        // depending on withdrawal queue status
+        int256 movedToWithdrawalBuffer = _processWithdrawals(
+            _requestIdToFinalizeUpTo,
+            _finalizationPooledEtherAmount,
+            _finalizationSharesAmount,
+            _wcBufferedEther);
+
+        // Update deposit buffer state
+        if (executionLayerRewards != 0 || movedToWithdrawalBuffer != 0) {
+            if (movedToWithdrawalBuffer > 0) {
+                BUFFERED_ETHER_POSITION.setStorageUint256(
+                    _getBufferedEther().add(executionLayerRewards).add(uint256(movedToWithdrawalBuffer))
+                );
+            } else {
+                BUFFERED_ETHER_POSITION.setStorageUint256(
+                    _getBufferedEther().add(executionLayerRewards).sub(uint256(-movedToWithdrawalBuffer))
+                );
+            }
+        }
+
+        return executionLayerRewards;
+    }
+
+    function _processRewards(
+        uint256 _beaconBalanceNew,
+        uint256 _executionLayerRewards,
+        uint256 _wcBufferedEther,
+        uint256 _beaconBalanceOld,
+        uint256 _appearedValidators
+    ) internal {
+        // Post-withdrawal rewards
+        // rewards = (beacon balance new - beacon balance old) - (appeared validators x 32 ETH)
+        // + withdrawn from execution layer rewards vault + withdrawn from withdrawal credentials vault
+
+        uint256 rewardsBase = (_appearedValidators.mul(DEPOSIT_SIZE)).add(_beaconBalanceOld);
+
+        // Don’t mint/distribute any protocol fee on the non-profitable Lido oracle report
+        // (when beacon chain balance delta is zero or negative).
+        // See ADR #3 for details:
+        // https://research.lido.fi/t/rewards-distribution-after-the-merge-architecture-decision-record/1535
+        if (_beaconBalanceNew.add(_wcBufferedEther) > rewardsBase) {
+            uint256 consensusLayerRewards = _beaconBalanceNew.add(_wcBufferedEther).sub(rewardsBase);
+            _distributeFee(consensusLayerRewards.add(_executionLayerRewards));
+        }
+    }
+
+    /**
+     * @dev finalize requests in the queue, burn shares and restake some ether if remains
+     * @return withdrawalFundsMovement amount of funds restaked (if positive) or moved to withdrawal buffer (if negative)
+     */
+    function _processWithdrawals(
+        uint256[] _requestIdToFinalizeUpTo,
+        uint256[] _finalizationPooledEtherAmount,
+        uint256[] _finalizationSharesAmount,
+        uint256 _wcBufferedEther
+    ) internal returns (int256 withdrawalFundsMovement) {
+        address withdrawalAddress = _getWithdrawalVaultAddress();
+        // do nothing if withdrawals is not configured
+        if (withdrawalAddress == address(0)) {
+            return 0;
+        }
+
+        IWithdrawalQueue withdrawal = IWithdrawalQueue(withdrawalAddress);
+
+        uint256 lockedEtherAccumulator = 0;
+
+        for (uint256 i = 0; i < _requestIdToFinalizeUpTo.length; i++) {
```

Comment:
Looks like all storage-access ops are performed within a loop.

Maybe worth move out them (I know that currently every oracle report is supposed to invoke only a single iteration, though, may change in the future).

Reply:
Yep, it's naive implementation and subject to further optimisation.

Reply:
Rechecked it and didn't get which ops can be moved out exactly. Can you elaborate?

Reply:
Ok, the major consideration was about `burnShares`.
Other possible directions:
-  require(lastIdToFinalize >= withdrawal.finalizedRequestsCounter(), "BAD_FINALIZATION_PARAMS") (don't call each iteration)
- withdrawal.calculateFinalizationParams (accept vector params)
- withdrawal.finalize.value(transferToWithdrawalBuffer) (accept vector params)